src/pretix/presale/checkoutflow.py

```diff
@@ -528,6 +529,40 @@ def provider_forms(self):
 
     def post(self, request):
         self.request = request
+        if request.POST.get("giftcard") and request.POST.get("payment") == "giftcard":
+            # TODO: cross-organizer acceptance, …
+            try:
+                gc = request.organizer.accepted_gift_cards.get(
+                    secret=request.POST.get("giftcard")
+                )
+                if gc.currency != request.event.currency:
+                    messages.error(self.request, _("This gift card does not support this currency."))
+                    return self.render()
+                if gc.pk in self.cart_session['gift_cards']:
+                    messages.error(self.request, _("This gift card is already used for your payment."))
+                    return self.render()
+                if gc.value <= Decimal("0.00"):
+                    messages.error(self.request, _("All credit on this gift card has been used."))
+                    return self.render()
+                if 'gift_cards' not in self.cart_session:
```

Comment:
This check needs to be moved up. I'm getting a key error in line 541. :D

Reply:
Silly me ;)